tests/common/mac/MacTestMain.cs

```diff
@@ -9,47 +10,85 @@
 using GuiUnit;
 using NUnit.Framework;
 
+// This is a bit of a hack. A number of BCL types have using statements with this namespace but no uses that we care about
+namespace NUnit.Framework.SyntaxHelpers
+{
+}
+
 namespace Xamarin.Mac.Tests
-{	
+{
 	static class MainClass
 	{
-		static void Main(string[] args)
+		static void Main (string[] args)
 		{
+#if !NO_GUI_TESTING
 			NSApplication.Init();
-			NSRunLoop.Main.InvokeOnMainThread(RunTests);
-			NSApplication.Main(args);
+#endif
+			RunTests ();
 		}
 
-		static void RunTests()
+		static void RunTests ()
 		{
-			TestRunner.MainLoop = new NSRunLoopIntegration();
+			TestRunner.MainLoop = new NSRunLoopIntegration ();
+			List<string> args = new List<string> () { typeof (MainClass).Assembly.Location, "-labels", "-noheader" };
+
 			string testName = System.Environment.GetEnvironmentVariable ("XM_TEST_NAME");
-			string [] args = testName != null ?
-				new [] { typeof(MainClass).Assembly.Location, "-labels", "-noheader", string.Format ("-test={0}", testName) } :
-				new [] { typeof(MainClass).Assembly.Location, "-labels", "-noheader" };
+			if (testName != null)
+				args.Add ($"-test={testName}");
 
-			TestRunner.Main (args);
+			if (System.Environment.GetEnvironmentVariable ("XM_BCL_TEST") != null)
+				args.Add ("-exclude=MobileNotWorking,NotOnMac,NotWorking,ValueAdd,CAS,InetAccess,NotWorkingInterpreter");
+			TestRunner.Main (args.ToArray ());
+#if NO_GUI_TESTING
+			while (true) {
+				System.Threading.Thread.Sleep (1000);
```

Comment:
This is a total hack. TestRunner.Main assumes you have a message pump spinning, but when I hack one out it returns right away.

We exit it via Environment.Exit farther down.